src/main/scala/io/iohk/ethereum/network/PeersInfoHolderActor.scala

```diff
@@ -0,0 +1,189 @@
+import io.iohk.ethereum.db.storage.AppStateStorage
+import io.iohk.ethereum.network.PeerActor.{DisconnectPeer, SendMessage}
+import io.iohk.ethereum.network.PeersInfoHolderActor._
+import io.iohk.ethereum.network.PeerEventBusActor.{PeerSelector, Subscribe, Unsubscribe}
+import io.iohk.ethereum.network.PeerEventBusActor.PeerEvent._
+import io.iohk.ethereum.network.PeerEventBusActor.SubscriptionClassifier._
+import io.iohk.ethereum.network.handshaker.Handshaker.HandshakeResult
+import io.iohk.ethereum.network.p2p.Message
+import io.iohk.ethereum.network.p2p.messages.CommonMessages.{NewBlock, Status}
+import io.iohk.ethereum.network.p2p.messages.PV62.{BlockHeaders, GetBlockHeaders, NewBlockHashes}
+import io.iohk.ethereum.network.p2p.messages.WireProtocol.Disconnect
+
+class PeersInfoHolderActor(peerEventBusActor: ActorRef, appStateStorage: AppStateStorage,
+                           forkResolverOpt: Option[ForkResolver]) extends Actor with ActorLogging {
+
+  val msgCodesWithInfo: Set[Int] = Set(BlockHeaders.code, NewBlock.code, NewBlockHashes.code)
+
+  peerEventBusActor ! Subscribe(PeerHandshaked)
+
+  override def receive: Receive = handlePeersInfoEvents(Map.empty)
+
+  def handlePeersInfoEvents(peersWithInfo: Map[PeerId, PeerWithInfo]): Receive = {
+
+    case MessageToPeer(message, peerId) if peersWithInfo.contains(peerId) =>
+      val PeerWithInfo(peer, oldPeerInfo) = peersWithInfo(peerId)
+      val newPeerInfo = handleSentMessage(message, oldPeerInfo)
+      context become handlePeersInfoEvents(peersWithInfo + (peerId -> PeerWithInfo(peer, newPeerInfo)))
+
+    case MessageFromPeer(message, peerId) if peersWithInfo.contains(peerId) =>
+      val PeerWithInfo(peer, oldPeerInfo) = peersWithInfo(peerId)
+      val newPeerInfo = handleReceivedMessage(message, oldPeerInfo, peer)
+      context become handlePeersInfoEvents(peersWithInfo + (peerId -> PeerWithInfo(peer, newPeerInfo)))
+
+    case PeerHandshakeSuccessful(peer, peerInfo: PeerInfo) =>
+      peerEventBusActor ! Subscribe(PeerDisconnectedClassifier(PeerSelector.WithId(peer.id)))
+      peerEventBusActor ! Subscribe(MessageReceivedClassifier(msgCodesWithInfo, PeerSelector.WithId(peer.id)))
+      peerEventBusActor ! Subscribe(MessageSentClassifier(msgCodesWithInfo, PeerSelector.WithId(peer.id)))
+
+      //Ask for the highest block from the peer
+      peer.ref ! SendMessage(GetBlockHeaders(Right(peerInfo.remoteStatus.bestHash), 1, 0, false))
+      context become handlePeersInfoEvents(peersWithInfo + (peer.id -> PeerWithInfo(peer, peerInfo)))
+
+    case PeerDisconnected(peerId) if peersWithInfo.contains(peerId) =>
+      peerEventBusActor ! Unsubscribe(PeerDisconnectedClassifier(PeerSelector.WithId(peerId)))
+      peerEventBusActor ! Unsubscribe(MessageReceivedClassifier(msgCodesWithInfo, PeerSelector.WithId(peerId)))
+      peerEventBusActor ! Unsubscribe(MessageSentClassifier(msgCodesWithInfo, PeerSelector.WithId(peerId)))
```

Comment:
If `MessageReceivedClassifier` and `MessageSentClassifier` have the same info and you are subscribing to both of them each time, what's the purpose of having two classes? Is there any case where we subscribe to one and not to the other and that's why we require to have them separated?